jlm/mlir/backend/mlirgen.cpp

```diff
@@ -0,0 +1,269 @@
+    convertTheta(theta, block);
+  } else if (auto delta = dynamic_cast<llvm::delta::node *>(node)) {
+    convertDelta(delta, block);
+  } else if (auto phi = dynamic_cast<llvm::phi::node *>(node)) {
+    convertPhi(phi, block);
+  */
+  } else {
+    throw util::error("Unimplemented structural node: " +
+		     node->operation().debug_string());
+  }
+}
+
+mlir::Value
+MLIRGen::convertSimpleNode(const rvsdg::simple_node *node, mlir::Block *block)
+{
+  if (auto bitsOp = dynamic_cast<const rvsdg::bitconstant_op *>(&(node->operation()))) {
+    auto constOp = builder->create<mlir::arith::ConstantIntOp>(builder->getUnknownLoc(), 1, 32);
+    block->push_back(constOp);
+    return constOp;
+  } else {
+    throw util::error("Unimplemented simple node: " +
+		     node->operation().debug_string());
+  }
+}
+
+mlir::Value
+MLIRGen::convertLambda(const llvm::lambda::node *lambdaNode, mlir::Block *block)
+{
+
+  // Handle function arguments
+  ::llvm::SmallVector<mlir::Type> arguments;
+  for (size_t i = 0; i < lambdaNode->nfctarguments(); ++i) {
+    arguments.push_back(convertType(&lambdaNode->fctargument(i)->type()));
+  }
+  ::llvm::ArrayRef argumentsArray(arguments);
+
+  // Handle function results
+  ::llvm::SmallVector<mlir::Type> results;
+  for (size_t i = 0; i < lambdaNode->nfctresults(); ++i) {
+      results.push_back(convertType(&lambdaNode->fctresult(i)->type()));
+  }
+  ::llvm::ArrayRef resultsArray(results);
+
+/*
+  // Context arguments
+  for (size_t i = 0; i < node->ncvarguments(); ++i) {
+    // s << print_input_origin(node.cvargument(i)->input()) << ": " << print_type(&ln.cvargument(i)->type());
+    throw util::error("Context arguments in convertLambda() has not been implemented");
+  }
+*/
+
+  // LambdaNodes return a LambdaRefType
+  ::llvm::SmallVector<mlir::Type> lambdaRef;
+  auto refType = builder->getType<::mlir::rvsdg::LambdaRefType>(argumentsArray, resultsArray);
+  lambdaRef.push_back(refType);
+
+  // Create the lambda node and add it to the region/block it resides in
+  ::llvm::SmallVector<mlir::Value> inputs;
+  auto lambda = builder->create<mlir::rvsdg::LambdaNode>(builder->getUnknownLoc(), lambdaRef, inputs);
+  block->push_back(lambda);
+
+  // Create a block for the region as this is not done automatically
+  mlir::Region &region = lambda.getRegion();
+  mlir::Block *lambdaBlock = new mlir::Block;
+  region.push_back(lambdaBlock);
+
+  // Convert the region and get all the results generated by the region
+  auto regionResults = convertSubregion(lambdaNode->subregion(), lambdaBlock);
+  auto lambdaResult = builder->create<mlir::rvsdg::LambdaResult>(builder->getUnknownLoc(), *regionResults);
+  lambdaBlock->push_back(lambdaResult);
+
+  return lambda;
+}
+
+mlir::Type
+MLIRGen::convertType(const rvsdg::type *type)
+{
+  if(auto bt = dynamic_cast<const rvsdg::bittype *>(type)){
+    return builder->getIntegerType(bt->nbits());
+  } else if (auto loopstateType = dynamic_cast<const llvm::loopstatetype*>(type)){
```

Comment:
You can remove `auto loopstateType =`, it is nowhere used.

Reply:
Removed